### PR TITLE
Remove the stored context object in MediaCoordinator and MediaThumbnailCoordinator

### PR DESCRIPTION
**Pull request by crazytonyli**
## Changes

The main changes in this PR falls into two categories:
1. Stop using the stored context object, replace it with the one provided by `performAndSave` function. This of course introduced a side-effect of all functions which used to share the serial queue inside the stored context object now doesn't execute sequentially. I've looked into each function and don't see how this change would introduce a negative impact. Except one case:
2. `syncMediaLibrary` is guarded using a serial `OperationQueue`, to make sure only one syncing is running at any given time.

## Test instructions

Try a few things around media library, i.e.
- Install a fresh copy of the app, go to Media screen to sync the media library
- Add a few photos to the media library or a post.
- Delete a few photos from the media library or a post.
- Log into WP.com from a web browser, add a few photos there and use the app to sync them.

------

[Hiding whitespace](https://github.com/wordpress-mobile/WordPress-iOS/pull/19538/files?w=1) might help code review.

Please note, changes in this PR are added on top of https://github.com/wordpress-mobile/WordPress-iOS/pull/19521, I'll rebase this PR's commits once #19521 is merged.

- [x] Switch to trunk branch.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
None.

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Services/MediaThumbnailCoordinator.swift [hunk 1]
```diff
@@ -53,11 +41,18 @@ class MediaThumbnailCoordinator: NSObject {
...
             DispatchQueue.main.async {
                 onCompletion(nil, error)
             }
-        })
+        }
+
+        coreDataStack.performAndSave { context in
+            let mediaThumbnailService = MediaThumbnailService(managedObjectContext: context)
+            mediaThumbnailService.exportQueue = self.queue
+            mediaThumbnailService.thumbnailURL(forMedia: media, preferredSize: size, onCompletion: success, onError: failure)
```
Comment (mokagio) [inline WordPress/Classes/Services/MediaThumbnailCoordinator.swift:54]: Was there a particular threading or memory reason for extracting `success` and `failure` as opposed to define them as closure parameters in the method call, or just style?

**Review by mokagio — APPROVED** [review]
Code looks great.

I haven't had a chance to test following the instructions and I don't know if I'll get to it soon. Leaving an approval in meantime to unblock.

File: WordPress/Classes/Services/MediaThumbnailCoordinator.swift (hunk 1 above)
Comment (crazytonyli) [reply WordPress/Classes/Services/MediaThumbnailCoordinator.swift:54]: Just style. I wanted to minimize the diff mainly. 😄

Comment (mokagio) [reply WordPress/Classes/Services/MediaThumbnailCoordinator.swift:54]: Sounds good. Merge at will 👍

**Pull request merged**